create a new external notifications service

nick-funk opened the pull request:
## What does this PR do?

allows events to be forwarded to an external notifications service for external consumption

## These changes will impact:

- [ ] commenters
- [ ] moderators
- [ ] admins
- [x] developers

## What changes to the GraphQL/Database Schema does this PR introduce?

None

## Does this PR introduce any new environment variables or feature flags?

- `EXTERNAL_NOTIFICATIONS_API_URL`
- `EXTERNAL_NOTIFICATIONS_API_KEY`

## If any indexes were added, were they added to `INDEXES.md`?

N/A

## How do I test this PR?

- set the `EXTERNAL_NOTIFICATIONS_API_URL` to the api url you want to forward notifications to
- set the `EXTERNAL_NOTIFICATIONS_API_KEY` to the value of the api key you will use with your external notifications service
- reply and rec content
- see that notifications calls are made to your external service

## Were any tests migrated to React Testing Library?

No

## How do we deploy this PR?

- merge into `develop`
- release with other changes via `develop` into `main`

nick-funk committed:
create a new external notifications service

allows events to be forwarded to an external
notifications service for external consumption

kabeaty commented on server/src/core/server/services/comments/actions.ts:
```diff
@@ -353,6 +355,20 @@ export async function createReaction(
     });
+
+    if (externalNotifications.active()) {
+      const parentAuthor = comment.authorID
```
Using `parentAuthor` and `author` here for recs doesn't seem as clear to me as it does for reply notifications. What about something more like `commentAuthor` and `recAuthor` or `recUser` or `userWhoRecced`?

nick-funk replied:
great idea, updated!

kabeaty replied:
nice, thanks for updating!

kabeaty commented on server/src/core/server/config.ts:
```diff
@@ -505,6 +505,18 @@ const config = convict({
+  external_notifications_api_key: {
+    doc: "API key to use when forwarding notifications to an external url",
+    format: String,
+    default: null,
```
I think this needs to have a default of empty string instead of null, as otherwise there is an error when starting up the server if nothing is set for the external notifications api key. Same note for the api key; it also seems to need a default url.

I guess then the check in the notifications service to see if it's active needs to be updated too to check for not empty string, etc.

nick-funk replied:
it turns out that empty strings are falsy, so that works for us

updated and fixed!

kabeaty replied:
The `external_notifications_api_key` needs to have a valid default url though, not just empty string. When starting up without anything set, it gives this error now:
```
 "message": "external_notifications_api_url: must be a URL",
```

nick-funk replied:
I've set this to the default url of `http://localhost:7003/api`, that should stop the errors when this isn't set in the env vars

this also works for the active state because you also have to set a valid API key for the service to become active 🙂

kabeaty commented on server/src/core/server/services/notifications/externalService.ts:
```diff
@@ -0,0 +1,188 @@
+      },
+    };
+
+    const response = await fetch(this.url!, {
```
I'm getting 
```
"message": "INTERNAL_ERROR: fetch is not defined",
```

I think this needs to be updated to use `import { createFetch } from "coral-server/services/fetch";` to create fetch here to use?

nick-funk replied:
oh geeze, nah, it's cause we need to use `node-fetch` in node v18 still, I was running on v20 for the vscode config, repairing now!

nick-funk replied:
fixed

kabeaty reviewed (commented):
This is looking good overall! I left a few comments throughout where I am seeing errors while testing this out.

nick-funk committed:
use direct import for `convert` in `htmlToText` lib

nick-funk committed:
import fetch in `externalNotifications`

- still supporting our ancient node 18 :sweat_smile:

nick-funk committed:
set default empty string for external notifications config values

nick-funk committed:
clarify external notifications from/to user properties

- use clear naming of
 - `reccingUser` and `reccedUser`
 - `replyingUser` and `repliedToUser`

tessalt commented on server/src/core/server/stacks/createComment.ts:
```diff
@@ -443,6 +445,24 @@ export default async function create(
         type,
       });
+
+      // if we have an external notifications service hooked up
```
should we skip doing default notification if the external service is enabled?

nick-funk replied:
good call, this is now hooked up to disable internal notifications when the external service is active

kabeaty reviewed (changes requested):
Thank you for the updates! This looks and works great overall now. Once the comment is addressed about providing a default url for `external_notifications_api_key` so that there is no error on server startup if it's not set, should be good to go.

nick-funk committed:
default external notifications api url to localhost

nick-funk committed:
handle errors gracefully within the external notifications service

nick-funk committed:
disable internal notifications when external notifications are active

nick-funk committed:
append `/api` to default external notifications url

nick-funk committed:
Merge branch 'develop' into feat/external-notifications-service

kabeaty reviewed (approved):
looking good!
